### DISCUSSION TITLE
Is there a way to print inbound/outbound connections to other peers?

### DISCUSSION THREAD

**Discussion by lynic** [libp2p/rust-libp2p#3025] [Q&A]
Hi, once I start a peer with DHT, is there a way to print all inbound/outbound connections to which peer on which multiaddr?
Also is there a way to show the details of sub streams, which Networkbehavior holds it ?

**Comment by mxinden** [comment]
> Hi, once I start a peer with DHT, is there a way to print all inbound/outbound connections to which peer on which multiaddr?

I suggest taking a loot at [`SwarmEvent`](https://docs.rs/libp2p/latest/libp2p/swarm/enum.SwarmEvent.html) emitted by `Swarm`.

> Hi, once I start a peer with DHT, is there a way to print all inbound/outbound connections to which peer on which multiaddr? Also is there a way to show the details of sub streams, which Networkbehavior holds it ?

Nit pick, it is a `ConnectionHandler` that holds a substream.

For now there are only logs. I could see ourself emitting an event after each succesful `multistream-select negotiation. See also https://github.com/libp2p/rust-libp2p/issues/2865.

**Reply by lynic** [reply]
Thanks , yes, now I watch for ConnectionEstablished and ConnectionClosed events and maintains a structs to record the inbound/outbound connections. Just wondering is there a way to get this info from swarm.

**Reply by thomaseizinger** [reply]
A more encapsulated way would be to implement a `NetworkBehaviour` and override the correct callbacks. That way you have less work to do when polling the swarm.
